src/core/audio/AudioFile.cpp

```diff
@@ -0,0 +1,184 @@
+    auto sfFormat = 0;
+    auto sfBitDepth = 0;
+    auto sfChannels = 0;
+
+    switch (settings.getBitDepth())
+    {
+	case OutputSettings::BitDepth::Depth16Bit:
+        sfBitDepth = SF_FORMAT_PCM_16;
+        break;
+	case OutputSettings::BitDepth::Depth24Bit:
+        sfBitDepth = SF_FORMAT_PCM_24;
+        break;
+	case OutputSettings::BitDepth::Depth32Bit:
+        sfBitDepth = SF_FORMAT_FLOAT;
+		break;
+	}
+
+	switch (format)
+    {
+	case AudioFileFormat::WAV:
+        sfFormat = SF_FORMAT_WAV | sfBitDepth;
+        break;
+	case AudioFileFormat::FLAC:
+        sfFormat = SF_FORMAT_FLAC | sfBitDepth;
+        break;
+	case AudioFileFormat::OGG:
+        sfFormat = SF_FORMAT_OGG | SF_FORMAT_VORBIS;
+        break;
+	case AudioFileFormat::MP3:
+        sfFormat = SF_FORMAT_MPEG | SF_FORMAT_MPEG_LAYER_III;
+        break;
+    default:
+        return SF_INFO{};
+	}
+
+    switch (settings.getStereoMode())
+    {
+	case OutputSettings::StereoMode::Stereo:
+	case OutputSettings::StereoMode::JointStereo:
+        sfChannels = 2;
+        break;
+	case OutputSettings::StereoMode::Mono:
+        sfChannels = 1;
+		break;
+	}
+
+	return {.samplerate = static_cast<int>(settings.getSampleRate()), .channels = sfChannels, .format = sfFormat};
+}
+
+} // namespace
+
+namespace lmms {
+struct AudioFile::Impl
+{
+	Impl(std::filesystem::path path);
+	Impl(std::filesystem::path path, AudioFileFormat format, OutputSettings settings);
+	SF_INFO m_info;
+	SNDFILE* m_sndfile;
+	std::filesystem::path m_path;
+};
+
+AudioFile::AudioFile(std::filesystem::path path)
+	: m_impl(std::make_unique<Impl>(path))
+{
+	if (!m_impl->m_sndfile)
+	{
+		throw std::runtime_error{"failed to construct audio file: " + std::string{sf_strerror(m_impl->m_sndfile)}};
+	}
+}
+
+AudioFile::AudioFile(std::filesystem::path path, AudioFileFormat format, OutputSettings settings)
+	: m_impl(std::make_unique<Impl>(path, format, settings))
+{
+    if (!m_impl->m_sndfile)
+	{
+		throw std::runtime_error{"failed to construct audio file: " + std::string{sf_strerror(m_impl->m_sndfile)}};
+	}
+}
+
+AudioFile::~AudioFile() = default;
+
+AudioFile::Impl::Impl(std::filesystem::path path)
+	: m_info()
+	, m_sndfile(openAudioFile(path, SFM_READ, &m_info))
+	, m_path(path)
+{
+}
+
+AudioFile::Impl::Impl(std::filesystem::path path, AudioFileFormat format, OutputSettings settings)
+	: m_info(sfInfoFromOutputSettings(format, settings))
+	, m_sndfile(openAudioFile(path, SFM_WRITE, &m_info))
+	, m_path(path)
+{
+	if (format == AudioFileFormat::FLAC)
+	{
+		auto compressionLevel = settings.getCompressionLevel();
+		sf_command(m_sndfile, SFC_SET_COMPRESSION_LEVEL, &compressionLevel, sizeof(int));
+	}
+	else if (format == AudioFileFormat::MP3 || format == AudioFileFormat::OGG)
+	{
+		constexpr auto minBitRate = 32;
+		constexpr auto maxBitRate = 320;
+		const auto targetBitRate = settings.bitrate();
+
+		auto compressionLevel = (maxBitRate - targetBitRate) / static_cast<double>(maxBitRate - minBitRate);
```

Comment:
This is correct for MP3, but OGG Vorbis uses somewhat different values.
Per data from `lib/modes` in libvorbis:
```cpp
// Stereo, per-channel (multiply by 2 for a full file)
static const double rate_mapping_44_stereo[12]={
  22500.,32000.,40000.,48000.,56000.,64000.,
  80000.,96000.,112000.,128000.,160000.,250001.
};

// Mono
static const double rate_mapping_44_un[12]={
  32000.,48000.,60000.,70000.,80000.,86000.,
  96000.,110000.,120000.,140000.,160000.,240001.
};
```
Note that the values correspond to quality level -0.1 to 1.0, where the compression level is (1 - quality_level) with range limited by `[0, 1]`.